File: frontend/src/types/practice.ts
import type { Difficulty } from './roadmap'

export type QuestionType = 'mcq' | 'msq' | 'nat'
export type QuestionSource = 'pyq' | 'practice' | 'ai_generated' | 'mock'

export interface QuestionPublic {
  id: string
  topic_id: string
  question_type: QuestionType
  difficulty: Difficulty
  source: QuestionSource
  source_year: number | null
  question_markdown: string
  options: string[] | null
  marks: number
  negative_marks: number
  expected_time_seconds: number
}

export interface PracticeStartRequest {
  topic_id?: string
  subject_id?: string
  difficulty?: Difficulty
  question_count: number
}

export interface PracticeAttemptStarted {
  attempt_id: string
  topic_id: string | null
  questions: QuestionPublic[]
}

export interface PracticeAnswerRequest {
  question_id: string
  selected_indices?: number[]
  nat_value?: number
  is_skipped: boolean
  time_taken_seconds: number
}

export interface PracticeAnswerResult {
  question_id: string
  is_correct: boolean | null
  marks_awarded: number
  correct_option_indices: number[]
  correct_value: number | null
  explanation_markdown: string | null
}

export interface PracticeSubmitResponse {
  attempt_id: string
  score: number
  max_score: number
  accuracy_percent: number
  correct_count: number
  incorrect_count: number
  skipped_count: number
  negative_marks_lost: number
  avg_time_per_question_seconds: number
}

export interface TopicQuestionStats {
  total: number
  by_difficulty: Record<string, number>
  by_type: Record<string, number>
}

File: backend/app/schemas/practice.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import Difficulty, QuestionSource, QuestionType


class PracticeStartRequest(BaseModel):
    topic_id: int | None = None
    subject_id: int | None = None
    difficulty: Difficulty | None = None
    question_count: int = 10


class QuestionPublic(BaseModel):
    """Question as shown DURING an attempt — never includes the answer."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    topic_id: int
    question_type: QuestionType
    difficulty: Difficulty
    source: QuestionSource
    source_year: int | None
    question_markdown: str
    options: list[str] | None
    marks: float
    negative_marks: float
    expected_time_seconds: int


class PracticeAttemptStarted(BaseModel):
    attempt_id: int
    topic_id: int | None
    questions: list[QuestionPublic]


class PracticeAnswerRequest(BaseModel):
    question_id: int
    selected_indices: list[int] | None = None
    nat_value: float | None = None
    is_skipped: bool = False
    time_taken_seconds: int


class PracticeAnswerResult(BaseModel):
    question_id: int
    is_correct: bool | None
    marks_awarded: float
    correct_option_indices: list[int]
    correct_value: float | None
    explanation_markdown: str | None


class PracticeSubmitResponse(BaseModel):
    attempt_id: int
    score: float
    max_score: float
    accuracy_percent: float
    correct_count: int
    incorrect_count: int
    skipped_count: int
    negative_marks_lost: float
    avg_time_per_question_seconds: float


class PracticeAttemptSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    topic_id: int | None
    started_at: datetime
    submitted_at: datetime | None
    score: float | None
    accuracy_percent: float | None
    is_complete: bool


class TopicQuestionStats(BaseModel):
    total: int
    by_difficulty: dict[str, int]
    by_type: dict[str, int]
